### PR TITLE
Complete separating of packages

### PR DESCRIPTION
**Pull request by ArlonAntonius**
Change Log:
- Create Affects Connection, responsible for setting a special tenant connection
- Create Hooks Database (Mutation), responsible for creating, updating and deleting of databases
- Removed ProvidesPassword & PasswordGenerator, ! NEEDS ATTENTION !
- Remove all Connection / Database related code in the Tenancy Framework
- Adjust Affects Models, does not depend on Affects Connection anymore and is more a general mutator now
- Removed Lifecycle Hooks Trait
- Added Traits for separated packages that provide better support for testing

A few things:
- I think we need to discuss what we want to do with the ProvidesPassword, should it come back? If so, in the framework or as a separate package?
- Do we want to provide "helper" function to Affects Models so it's easier to adjust the connection of models?

### CODE REVIEW

**Comment by luceos** [comment]
Can you merge master into this branch and fix merge conflicts?

File: src/Affects/Connection/Events/Identified.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * @var string|null
+     */
+    public $connection;
+
+    public function __construct(Tenant $tenant = null, string $connection = null, ProvidesConfiguration &$provider)
+    {
+        $this->tenant = $tenant;
+        $this->provider = &$provider;
```
Comment (luceos) [inline src/Affects/Connection/Events/Identified.php]: sorting is wrong, the $provider is the third argument, so it should be assigned third in the method as property

File: src/Affects/Connection/Provider.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            Listeners\SetConnection::class,
+        ],
+    ];
+
+    public function boot()
+    {
+        parent::boot();
+        Environment::macro('getTenantConnectionName', function () {
+            return config('tenancy.connection.tenant-connection-name') ?? 'tenant';
+        });
```
Comment (luceos) [inline src/Affects/Connection/Provider.php:55]: 🤤 sexy

File: src/Affects/Connection/resources/config/connection.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * @see https://tenancy.dev
+ * @see https://github.com/tenancy
+ */
+
+return [
+    /*
+     * The name of the tenant connection, tenancy will create this connection during runtime.
+     */
+    'tenant-connection-name' => env('TENANCY_TENANT_CONNECTION_NAME', 'tenant'),
```
Comment (luceos) [inline src/Affects/Connection/resources/config/connection.php:21]: I don't think we need this. Can't we move this to the event?

File: src/Affects/Models/Events/ConfigureModels.php [hunk 4]
```diff
@@ -29,53 +26,30 @@ class ConfigureModels
...
     {
-        $resolver = count($models) ? $this->getResolver($reset) : null;
-
         foreach ($models as $model) {
             if (!class_exists($model)) {
                 throw new InvalidArgumentException("$model does not exist");
             }
 
-            forward_static_call([$model, 'setConnectionResolver'], $resolver);
+            forward_static_call([$model, $function], ...$arguments);
```
Comment (luceos) [inline src/Affects/Models/Events/ConfigureModels.php]: I don't like this "wildcard" ability at all inside an event like this with such a weird method name.. I'm extremely curious why this is necessary at all.

File: src/Database/Mysql/Driver/Mysql.php [hunk 5]
```diff
@@ -63,7 +63,8 @@ public function update(Tenant $tenant): bool
         $tempTenant->{$tempTenant->getTenantKeyName()} = $tenant->getOriginal($tenant->getTenantKeyName());
         $originalTenant = Tenancy::getTenant();
 
-        Tenancy::setTenant($tempTenant);
+        $resolver = resolve(\Tenancy\Affects\Connection\ConnectionResolver::class);
+        $resolver->__invoke($tempTenant, Tenancy::getTenantConnectionName());
```
Comment (luceos) [inline src/Database/Mysql/Driver/Mysql.php]: Looks like this method needs some cleaning up. Move setting and retrieving the connection to a new method?

File: src/Hooks/Database/resources/config/database.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    /*
+     * Automatic tenant database handling.
+     */
+
+    'auto-create' => env('TENANCY_DATABASE_AUTO_CREATE', true),
+
+    'auto-update' => env('TENANCY_DATABASE_AUTO_UPDATE', true),
+
+    'auto-delete' => env('TENANCY_DATABASE_AUTO_DELETE', true),
+];
```
Comment (luceos) [inline src/Hooks/Database/resources/config/database.php:27]: Not another configuration file? Can't we use a static setter instead?

**Review by luceos — CHANGES_REQUESTED** [review]
Quickly went over it, will do a more extensive review through checkout once these edits are done.

File: src/Affects/Connection/resources/config/connection.php (hunk 3 above)
Comment (ArlonAntonius) [reply src/Affects/Connection/resources/config/connection.php:21]: How are we reliably going to get this in the env then?
I feel like a simple small config might do better than another event that would set this config.

File: src/Affects/Models/Events/ConfigureModels.php (hunk 4 above)
Comment (ArlonAntonius) [reply src/Affects/Models/Events/ConfigureModels.php]: I want it to be as flexible as possible if people want to use the event to run methods on models, they should be able to, why would we only allow them to change the connectionResolver?

File: src/Database/Mysql/Driver/Mysql.php (hunk 5 above)
Comment (ArlonAntonius) [reply src/Database/Mysql/Driver/Mysql.php]: Have been thinking about moving the functionality of grabbing the tables from:
Using a tenant to basically query SHOW TABLES for us.
To:
Using the system connection to grab the tables using a raw query SHOW TABLES.

In that way, the code won't do any weird fiddling in order to grab the tables.

File: src/Hooks/Database/resources/config/database.php (hunk 6 above)
Comment (ArlonAntonius) [reply src/Hooks/Database/resources/config/database.php:27]: Do we really want to put in the effort in avoiding a config file for something so small?

**Review by luceos — CHANGES_REQUESTED** [review]
InteractsWithConnections is a Trait in Affects/Connection/Support, but that same folder also holds a Trait directory. In addition this trait seems to be used only in tests, should it go into the testing package?

**Pull request merged**